### PR TITLE
KAFKA-4565: Separation of Internal and External traffic (KIP-103)

### PR DESCRIPTION
**Pull request by ijuma**

### CODE REVIEW

**Comment by ijuma** [comment]
To be done:
- [x]  Add more tests.
- [x]  Fix one test failure
- [ ] Namespacing of security configs (in a separate PR)

File: clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java [hunk 1]
```diff
@@ -105,10 +99,14 @@ public String toString() {
...
 
-        public EndPoint(String host, int port) {
+        public EndPoint(String host, int port, SecurityProtocol securityProtocol, ListenerName listenerName) {
             this.host = host;
             this.port = port;
+            this.securityProtocol = securityProtocol;
+            this.listenerName = listenerName;
         }
 
         @Override
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java:112]: Might be worth updating the `toString`.

File: core/src/main/scala/kafka/client/ClientUtils.scala [hunk 2]
```diff
@@ -105,34 +106,46 @@ object ClientUtils extends Logging{
...
+      }
+      connected = channel != null
+    }
+
+    channel
+  }
+
    /**
-    * Creates a blocking channel to a random broker
+    * Returns the first end point with the PLAINTEXT security protocol.
```
Comment (hachikuji) [inline core/src/main/scala/kafka/client/ClientUtils.scala]: nit: "Returns the first end point _from each broker_ with ..."

File: core/src/main/scala/kafka/cluster/Broker.scala [hunk 3]
```diff
@@ -20,6 +20,7 @@ package kafka.cluster
 import kafka.common.{BrokerEndPointNotAvailableException, BrokerNotAvailableException, KafkaException}
 import kafka.utils.Json
 import org.apache.kafka.common.Node
+import org.apache.kafka.common.network.ListenerName
 import org.apache.kafka.common.protocol.SecurityProtocol
 
 /**
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Broker.scala]: The comments below describing the JSON structure need updating.

File: core/src/main/scala/kafka/cluster/Broker.scala [hunk 4]
```diff
@@ -79,17 +80,21 @@ object Broker {
             else if (version == 1) {
               val host = brokerInfo("host").asInstanceOf[String]
               val port = brokerInfo("port").asInstanceOf[Int]
-              Map(SecurityProtocol.PLAINTEXT -> new EndPoint(host, port, SecurityProtocol.PLAINTEXT))
+              val securityProtocol = SecurityProtocol.PLAINTEXT
+              val endPoint = new EndPoint(host, port, ListenerName.forSecurityProtocol(securityProtocol), securityProtocol)
+              Seq(endPoint)
             }
             else {
+              val securityProtocolMap = brokerInfo.get("listener.security.protocol.map").map(
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Broker.scala]: nit: the indentation, lack of names and reuse of `map` here makes this a bit hard to follow. Would something like this be clearer?
```scala
val securityProtocolMap = brokerInfo.get("listener.security.protocol.map").map { configValue =>
  configValue.asInstanceOf[Map[String, String]].map { case (listenerName, securityProtocol) =>
    new ListenerName(listenerName) -> SecurityProtocol.forName(securityProtocol)
  }
}
```

File: clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java [hunk 5]
```diff
@@ -498,16 +499,22 @@ private UpdateMetadataRequest createUpdateMetadataRequest(int version, String ra
         partitionStates.put(new TopicPartition("topic20", 1),
                 new PartitionState(1, 0, 1, new ArrayList<>(isr), 2, new HashSet<>(replicas)));
 
-        Map<SecurityProtocol, UpdateMetadataRequest.EndPoint> endPoints1 = new HashMap<>();
-        endPoints1.put(SecurityProtocol.PLAINTEXT, new UpdateMetadataRequest.EndPoint("host1", 1223));
+        SecurityProtocol plaintext = SecurityProtocol.PLAINTEXT;
```
Comment (gwenshap) [inline clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:505]: Awkwardly placed comment: I think you are not actually testing V3...

i.e. I'd expect to see some test that calls createUpdateMetadataRequest(3,"rack") or something similar.

Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:505]: I know. :) I haven't pushed the new tests as those are still in progress (as I said in a PR comment that probably got lost in the Jenkins bot messages).

File: core/src/main/scala/kafka/cluster/Broker.scala (hunk 4 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/cluster/Broker.scala]: Mentioned offline, but the map is expecting underscores in the name. Maybe we could use a constant somewhere to avoid the potential for inconsistency?

File: clients/src/main/java/org/apache/kafka/common/network/ListenerName.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * Create an instance with the provided value converted to uppercase.
+     */
+    public static ListenerName normalised(String value) {
+        return new ListenerName(value.toUpperCase(Locale.ROOT));
+    }
+
+    private final String value;
+
+    public ListenerName(String value) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/network/ListenerName.java:42]: Should the name always be normalized? (In other words, should this be a private constructor?)

File: core/src/main/scala/kafka/server/MetadataCache.scala [hunk 7]
```diff
@@ -164,13 +164,13 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
       aliveNodes.clear()
       aliveBrokers.clear()
       updateMetadataRequest.liveBrokers.asScala.foreach { broker =>
-        val nodes = new EnumMap[SecurityProtocol, Node](classOf[SecurityProtocol])
-        val endPoints = new EnumMap[SecurityProtocol, EndPoint](classOf[SecurityProtocol])
-        broker.endPoints.asScala.foreach { case (protocol, ep) =>
-          endPoints.put(protocol, EndPoint(ep.host, ep.port, protocol))
-          nodes.put(protocol, new Node(broker.id, ep.host, ep.port))
+        val nodes = new java.util.HashMap[ListenerName, Node]
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/MetadataCache.scala:170]: Curious if there's any reason not to use a mutable scala map.

File: core/src/main/scala/kafka/cluster/Broker.scala [hunk 8]
```diff
@@ -79,17 +80,21 @@ object Broker {
             else if (version == 1) {
               val host = brokerInfo("host").asInstanceOf[String]
               val port = brokerInfo("port").asInstanceOf[Int]
-              Map(SecurityProtocol.PLAINTEXT -> new EndPoint(host, port, SecurityProtocol.PLAINTEXT))
+              val securityProtocol = SecurityProtocol.PLAINTEXT
+              val endPoint = new EndPoint(host, port, ListenerName.forSecurityProtocol(securityProtocol), securityProtocol)
+              Seq(endPoint)
             }
             else {
+              val securityProtocolMap = brokerInfo.get("listener_security_protocol_map").map(
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/Broker.scala]: Is there a convenient place to put a constant for this field? Also, the second `map` makes the lines below a bit tough to follow. Could we do this instead:
```scala
val securityProtocolMap = brokerInfo.get("listener.security.protocol.map").map(
  _.asInstanceOf[Map[String, String]].map{ case (listenerName, securityProtocol) =>
  new ListenerName(listenerName) -> SecurityProtocol.forName(securityProtocol)
})
```

**Review by hachikuji — COMMENTED** [review]
Left some minor comments. Will finish reviewing later today.

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 9]
```diff
@@ -407,6 +413,9 @@ object KafkaConfig {
   val AdvertisedListenersDoc = "Listeners to publish to ZooKeeper for clients to use, if different than the listeners above." +
   " In IaaS environments, this may need to be different from the interface to which the broker binds." +
   " If this is not set, the value for `listeners` will be used."
+  val ListenerSecurityProtocolMapDoc = "Mapping between listener names and security protocols. Key and value are separated by a colon " +
```
Comment (gwenshap) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Since this is the main documentation - I'd give a bit more background. "You may want to configure the same security protocol on two different ports or IPs - for example to separate external and internal traffic. To do that..."

File: clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java (hunk 5 above)
Comment (gwenshap) [reply clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:505]: Yeah, I did see your comment - you updated the tests just enough to get things to build and to confuse me a bit :)

File: core/src/main/scala/kafka/server/MetadataCache.scala (hunk 7 above)
Comment (ijuma) [reply core/src/main/scala/kafka/server/MetadataCache.scala:170]: We optimised this code quite a bit in a previous release as it affects the performance of metadata requests. The default Scala HashMap is a bit slower than Java's HashMap. In Scala 2.11, there's `AnyRefMap` that is comparable to Java's, but we still support 2.10.

File: clients/src/main/java/org/apache/kafka/common/network/ListenerName.java (hunk 6 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/network/ListenerName.java:42]: I was wondering about that. I decided to keep it like this because there are a number of situations where we know that the String has already been normalised, so we don't need to do it again. However, always normalising is safer so I am willing to change it if you feel strongly.

**Comment by gwenshap** [comment]
Read through the whole thing and didn't have many comments - I'm impressed at how clean this fairly large change ended up. You touch a lot of places, but most of it fairly mechanical replacement since the APIs match well. Nice :)

Since this is large, let me know if there are areas you want me to give a bit more attention.

File: core/src/main/scala/kafka/cluster/Broker.scala (hunk 8 above)
Comment (ijuma) [reply core/src/main/scala/kafka/cluster/Broker.scala]: Yeah, I added a note to do that along with the test that verifies the typo fix. I'll change the code to avoid the second map.

File: clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java (hunk 1 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java:112]: Yeah, I noticed that too. But a good way to test reviewer alertness. :)

File: clients/src/main/java/org/apache/kafka/common/network/ListenerName.java (hunk 6 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/network/ListenerName.java:42]: I'm ambivalent as well. Let's just keep it as it is.

**Comment by ijuma** [comment]
Thanks for the review @gwenshap! I'm particularly interested in feedback regarding the user-visible docs. I tried to address your suggestion and made a few other minor tweaks in the docs for listeners, advertised.listeners and server.properties. Let me know if you have suggestions for improvement.

**Comment by ijuma** [comment]
Started a system tests run:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/665/

The only remaining thing on my list is an integration test containing two listeners with the same security protocol. Working on that at the moment.

File: core/src/main/scala/kafka/cluster/EndPoint.scala [hunk 10]
```diff
@@ -17,63 +17,58 @@
...
       else
         Utils.formatAddress(host, port)
-    protocolType + "://" + hostport
-  }
-
-  def writeTo(buffer: ByteBuffer): Unit = {
-    buffer.putInt(port)
-    writeShortString(buffer, host)
-    buffer.putShort(protocolType.id)
+    listenerName.value() + "://" + hostport
```
Comment (hachikuji) [inline core/src/main/scala/kafka/cluster/EndPoint.scala]: nit: unneeded parenthesis

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 11]
```diff
@@ -690,8 +706,9 @@ object KafkaConfig {
       .define(LeaderImbalancePerBrokerPercentageProp, INT, Defaults.LeaderImbalancePerBrokerPercentage, HIGH, LeaderImbalancePerBrokerPercentageDoc)
       .define(LeaderImbalanceCheckIntervalSecondsProp, LONG, Defaults.LeaderImbalanceCheckIntervalSeconds, HIGH, LeaderImbalanceCheckIntervalSecondsDoc)
       .define(UncleanLeaderElectionEnableProp, BOOLEAN, Defaults.UncleanLeaderElectionEnable, HIGH, UncleanLeaderElectionEnableDoc)
-      .define(InterBrokerSecurityProtocolProp, STRING, Defaults.InterBrokerSecurityProtocol, MEDIUM, InterBrokerSecurityProtocolDoc)
+      .define(InterBrokerSecurityProtocolProp, STRING, null, MEDIUM, InterBrokerSecurityProtocolDoc)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: We document the default for this above as PLAINTEXT, but set it to null here. Would that cause confusion when we generate the ConfigDef table? Perhaps worth it to make config parsing easier.

File: core/src/main/scala/kafka/server/MetadataCache.scala (hunk 7 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/server/MetadataCache.scala:170]: Ah, yes, I remember. Might be worth a comment so no one is tempted to change it.

File: core/src/main/scala/kafka/utils/ZkUtils.scala [hunk 12]
```diff
@@ -360,25 +350,14 @@ class ZkUtils(val zkClient: ZkClient,
...
-                      "port" -> port,
-                      "endpoints" -> advertisedEndpoints.values.map(_.connectionString).toArray,
-                      "jmx_port" -> jmxPort,
-                      "timestamp" -> timestamp
-    )
-    rack.foreach(rack => if (version >= 3) jsonMap += ("rack" -> rack))
-
-    val brokerInfo = Json.encode(jsonMap)
-    registerBrokerInZk(brokerIdPath, brokerInfo)
+    val version = if (apiVersion >= KAFKA_0_10_0_IV1) 4 else 2
```
Comment (hachikuji) [inline core/src/main/scala/kafka/utils/ZkUtils.scala:359]: This doesn't look right. Shouldn't we use `KAFKA_0_10_2_IV0`?

**Review by hachikuji — COMMENTED** [review]
Looking good other than a question in `ZkUtils.registerBrokerInZk`.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 11 above)
Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaConfig.scala]: Yeah, it's not ideal, but I am not sure how to implement the desired behaviour otherwise: it should default to PLAINTEXT if the user hasn't set it, but it should give an error if the user has set `inter.broker.listener.name` and `security.inter.broker.protocol`. If we leave the `PLAINTEXT` default in the config definition, it seems like we cannot distinguish the case where the user set it versus the default. Hmm, maybe we can use `originals()` for this, I'll try it.

File: core/src/main/scala/kafka/utils/ZkUtils.scala (hunk 12 above)
Comment (ijuma) [reply core/src/main/scala/kafka/utils/ZkUtils.scala:359]: It looks weird, but it is actually right, I think. The method documentation explains it:

```java
* Register brokers with v4 json format (which includes multiple endpoints and rack) if
   * the apiVersion is 0.10.0.X or above. Register the broker with v2 json format otherwise.
   * Due to KAFKA-3100, 0.9.0.0 broker and old clients will break if JSON version is above 2.
   * We include v2 to make it possible for the broker to migrate from 0.9.0.0 to 0.10.0.X or above without having to
   * upgrade to 0.9.0.1 first (clients have to be upgraded to 0.9.0.1 in any case).
   *
   * This format also includes default endpoints for compatibility with older clients.
```

We only need to keep the version low for 0.9.0.0 brokers that had a bug in their parsing logic. More recent brokers will happily read newer versions as they expect no incompatible changes  in the format of the existing fields. I added  a short comment referencing the method docs near that code.

Does that make sense?

**Comment by ijuma** [comment]
System tests passed apart from the Streams system tests (both are failing in trunk as well).

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 11 above)
Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaConfig.scala]: OK, I pushed a change that solves the issue in a nicer way, I think.

**Comment by ijuma** [comment]
I pushed an integration test and I am now happy with this PR. Please take a look @hachikuji and we can merge it once you're happy.

File: core/src/main/scala/kafka/utils/ZkUtils.scala (hunk 12 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/utils/ZkUtils.scala:359]: Thanks, makes sense.

File: core/src/test/scala/integration/kafka/server/MultipleListenersWithSameSecurityProtocolTest.scala [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+  /**
+    * Tests that we can produce and consume to/from all broker-defined listeners and security protocols. We produce
+    * with acks=-1 to ensure that replication is also working.
+    */
+  @Test
+  def testProduceConsume(): Unit = {
+    producers.foreach { case (listenerName, producer) =>
+      val producerRecords = (1 to 10).map(i => new ProducerRecord(listenerName.value, s"key$i".getBytes,
+        s"value$i".getBytes))
+      producerRecords.map(producer.send(_)).map(_.get(10, TimeUnit.SECONDS))
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/server/MultipleListenersWithSameSecurityProtocolTest.scala:112]: nit: guess we could use `producer.send` instead of `producer.send(_)`

**Review by hachikuji — APPROVED** [review]
LGTM. Very clean implementation!
